resources/lib/actions/iptvmanageraction.py

```diff
@@ -0,0 +1,76 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import json
+import socket
+
+from resources.lib.actions.addonaction import AddonAction
+from resources.lib.logger import Logger
+from resources.lib.helpers.channelimporter import ChannelIndex
+
+
+class IPTVManagerAction(AddonAction):
+    def __init__(self, parameter_parser, request, port):
+        """ Sending steam and EPG data to IPTV Manager
+
+        :param ActionParser parameter_parser:  A ActionParser object to is used to parse
+                                                and create urls
+        :param string request:          The data to query [channels, epg]
+        :param integer port:            The port number to communicate with IPTV Manager
+
+        """
+
+        super(IPTVManagerAction, self).__init__(parameter_parser)
+
+        self.__request = request
+        self.__port = port
+        self.__parameter_parser = parameter_parser
+
+    def execute(self):
+        """ Send the output of the wrappers to socket """
+
+        Logger.debug("Execute IPTVManagerAction")
+        if self.__request == "streams":
+            self.send_streams()
+        if self.__request == "epg":
+            self.send_epg()
+
+
+    def via_socket(func):
+        """Send the output of the wrapped function to socket"""
+
+        def send(self):
+            """Decorator to send over a socket"""
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.connect(('127.0.0.1', self.__port))
+            try:
+                sock.sendall(json.dumps(func(self)).encode())
+            finally:
+                sock.close()
+
+        return send
+
+    @via_socket
+    def send_streams(self):
+        """Return JSON-STREAMS formatted python datastructure to IPTV Manager"""
+        streams = [] 
+        channels = ChannelIndex.get_register().get_channels()
+
+        for channel in channels:
+            fetched_channel = channel.get_channel();
+            if "create_iptv_streams" in dir(fetched_channel):
```

Comment:
This is going to be painfully slow, as much is happing if you create all channels. So don't do that. 

Better is to add an `hasEpg` to the `chn_.....json` file, e.g.: 

```json 
    {
      "guid": "ED8FDB9C-3870-11E2-B1C2-192A6288709B",
      "name": "NPO Start",
      "description": {
          "en": "Broadcast of the Dutch public broadcasters via www.npostart.nl.\n\nA [COLOR=yellow]NPO Start account[/COLOR] can be configured via the add-on settings. This makes the 'subscribed' shows available within Retrospect.",
        "nl": "Uitzendingen van de publieke zenders beschikbaar via www.npostart.nl.\n\nVia de add-on instellingen kan een [COLOR=yellow]NPO Start account[/COLOR] worden geconfigureerd zodat de 'gevolgde' programma's beschikbaar komen binnen Retrospect."
      },
      "icon": "npolarge.png",
      "category": "National",
      "channelcode": "uzgjson",
      "sortorder": 1,
      "language": "nl",
      "fanart": "nosfanart4.jpg",
      "poster": "npoposter.png",
      "message": "By using the NPO Start channel you are accepting the NPO Terms of Use: https://www.npostart.nl/npo-id-gebruiksvoorwaarden.",
      "hasEpg": true
    }
```

And also update the `plugin.video.retrospect/channels/plugin.video.retrospect.channel.json` accordingly, so the json files can be auto completed. 

Then update `ChannelInfo.from_json` to include the EPG bool. That way you can do the check on line 60 on the actual ChannelInfo object, without creating the channel. And only if there was EPG, then load it.